### PR TITLE
Enable to use regexp instead string path

### PR DESCRIPTION
**Pull request by yarax**
Hi!
I had to listen different ws paths in my code by mask, so I think it would be useful for others

### CODE REVIEW

**Comment by nkzawa** [comment]
I'm curious when you need this.

**Comment by 3rd-Eden** [comment]
Also, pull requests without tests are accepted.

**Comment by yarax** [comment]
We balance ws connections among servers, using nginx and key in path (/ws/room1 should go to server1, /ws/room2 to server2)  and in code I just want to know, that it it something like /ws/* comes

If you approve feature I'll make tests

**Comment by yarax** [comment]
Another case I'd like to pass some data on handshake in path
And the another reason why I ask pull request is no ability to patch `handleUpgrade` method with my own one, because of using global functions in module

**Comment by sequoiar** [comment]
@yarax it should be 

-    if (u && u.pathname !== this.options.path) return;
+    if (this.options.path instanceof RegExp) {
+      if (u && u.pathname && !u.pathname.match(this.options.path)) return;
+    } else {
+      if (u && u.pathname !== this.options.path) return;
+    }
+
   }

**Comment by yarax** [comment]
@sequoiar sure, thank you

**Comment by nkzawa** [comment]
Can't we use query parameters for that? I think it's more appropriate than using path.

Btw, it seems `path` option is used in other places too.

https://github.com/websockets/ws/blob/master/lib/WebSocketServer.js#L57

**Comment by yarax** [comment]
@nkzawa nginx `location` directive doesn't match query string at all. So it's impossible.
And the another point is that I shouldn't synchronise balancing logic and code. I just should know that the coming request is suitable for the certain server. But coming url path may occur different

**Comment by michaelmcmillan** [comment]
> Also, pull requests without tests are accepted. Especially when they introduce new functionality.

Did you mean "unaccepted"?

**Comment by auchenberg** [comment]
+1 for this FR.

**Comment by pukhalski** [comment]
+1 for this PR.

File: lib/WebSocketServer.js [hunk 1]
```diff
@@ -140,7 +140,12 @@ WebSocketServer.prototype.handleUpgrade = function(req, socket, upgradeHead, cb)
   // check for wrong path
   if (this.options.path) {
     var u = url.parse(req.url);
-    if (u && u.pathname !== this.options.path) return;
+    if (this.options.path instanceof RegExp) {
+      if (u && u.pathname && !u.pathname.match(this.options.path)) return;
```
Comment (lpinca) [inline lib/WebSocketServer.js]: Can you use `RegExp.prototype.test()` instead of `String.prototype.match()`?

**Comment by lpinca** [comment]
@yarax can you add a test?
Thank you.

**Comment by lpinca** [comment]
@yarax sorry I've seen #483 only now. That may be a better way to handle this.

**Comment by lpinca** [comment]
Closing in favor of #483.

**Pull request closed without merging**